rempi@1.1.0 %gcc@11.2.0 build fails: rempi_message_manager.h:53:3: error: 'string' does not name a type; did you mean 'stdin'?

### Steps to reproduce the issue

Reproducible using:
* `spack@develop` (b32cb5765c598e59f2f91d99807b4331f4eba569 from `Fri Aug 12 23:17:20 2022 -0700`)
* `ecpe4s/ubuntu22.04-runner-x86_64:2022-08-01`
* `%gcc@11.2.0`

Concrete spec: [rempi.spec.json.txt](https://github.com/spack/spack/files/9332531/rempi.spec.json.txt)

Concretization:
```
 -   rempi@1.1.0%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]      ^autoconf@2.69%gcc@11.2.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu22.04-x86_64
[+]          ^m4@1.4.19%gcc@11.2.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64
[+]              ^diffutils@3.8%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                  ^libiconv@1.16%gcc@11.2.0 libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]              ^libsigsegv@2.13%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]          ^perl@5.34.1%gcc@11.2.0+cpanm+shared+threads arch=linux-ubuntu22.04-x86_64
[+]              ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu22.04-x86_64
[+]              ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared arch=linux-ubuntu22.04-x86_64
[+]              ^gdbm@1.19%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                  ^readline@8.1.2%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                      ^ncurses@6.2%gcc@11.2.0~symlinks+termlib abi=none arch=linux-ubuntu22.04-x86_64
[+]                          ^pkgconf@1.8.0%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]              ^zlib@1.2.12%gcc@11.2.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu22.04-x86_64
[+]      ^automake@1.16.5%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]      ^libpciaccess@0.16%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]          ^libtool@2.4.7%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]          ^util-macros@1.19.3%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]      ^mpich@4.0.2%gcc@11.2.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~two_level_namespace~vci~verbs~wrapperrpath datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64
[+]          ^findutils@4.9.0%gcc@11.2.0 patches=440b954 arch=linux-ubuntu22.04-x86_64
[+]          ^hwloc@2.8.0%gcc@11.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm+shared arch=linux-ubuntu22.04-x86_64
[+]              ^libxml2@2.9.13%gcc@11.2.0~python arch=linux-ubuntu22.04-x86_64
[+]                  ^xz@5.2.5%gcc@11.2.0+pic libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]          ^libfabric@1.14.1%gcc@11.2.0~debug~kdreg fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu22.04-x86_64
[+]          ^yaksa@0.2%gcc@11.2.0~cuda~rocm arch=linux-ubuntu22.04-x86_64
[+]              ^python@3.8.13%gcc@11.2.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu22.04-x86_64
[+]                  ^expat@2.4.8%gcc@11.2.0+libbsd arch=linux-ubuntu22.04-x86_64
[+]                      ^libbsd@0.11.5%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                          ^libmd@1.0.4%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                  ^gettext@0.21%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu22.04-x86_64
[+]                      ^tar@1.34%gcc@11.2.0 zip=pigz arch=linux-ubuntu22.04-x86_64
[+]                          ^pigz@2.7%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                          ^zstd@1.5.2%gcc@11.2.0+programs compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]                  ^libffi@3.4.2%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                  ^openssl@1.1.1q%gcc@11.2.0~docs~shared certs=mozilla patches=3fdcf2d arch=linux-ubuntu22.04-x86_64
[+]                      ^ca-certificates-mozilla@2022-07-19%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
[+]                  ^sqlite@3.38.5%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu22.04-x86_64
[+]                  ^util-linux-uuid@2.37.4%gcc@11.2.0 arch=linux-ubuntu22.04-x86_64
```

Install error:
```
$> spack install -f ./rempi.spec.json
...
==> Installing rempi-1.1.0-eeyr3icczf6v2agskpafphuqe5txl7uz
==> No binary for rempi-1.1.0-eeyr3icczf6v2agskpafphuqe5txl7uz found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/4f/4fd94fca52311fd19dc04a32547841e6c1c1656b7999b2f76f537d6ec24efccc.tar.gz
==> No patches needed for rempi
==> rempi: Executing phase: 'autoreconf'
==> rempi: Executing phase: 'configure'
==> rempi: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

4 errors found in build log:
     275           |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     276     rempi_err.cpp:348:12: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
     277       348 |     return "Unknown error";
     278           |            ^~~~~~~~~~~~~~~
     279     In file included from rempi_recorder.h:48,
     280                      from rempi_recorder.cpp:31:
  >> 281     rempi_message_manager.h:53:3: error: 'string' does not name a type; did you mean 'stdin'?
     282        53 |   string to_string();
     283           |   ^~~~~~
     284           |   stdin
     285     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_event.lo -MD -MP -MF .deps/librempi_la-rempi_event.Tpo -c rempi_event.cpp  -fPI
             C -DPIC -o .libs/librempi_la-rempi_event.o
     286     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_mf.lo -MD -MP -MF .deps/librempi_la-rempi_mf.Tpo -c rempi_mf.cpp -o librempi_la
             -rempi_mf.o >/dev/null 2>&1
     287     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_config.lo -MD -MP -MF .deps/librempi_la-rempi_config.Tpo -c rempi_config.cpp -o
              librempi_la-rempi_config.o >/dev/null 2>&1

     ...

     801     rempi_event.cpp:180:10: warning: converting to non-pointer type 'MPI_Request' {aka 'int'} from NULL [-Wconversion-null]
     802       180 |   return NULL;
     803           |          ^~~~
     804     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_clock.lo -MD -MP -MF .deps/librempi_la-rempi_clock.Tpo -c rempi_clock.cpp -o li
             brempi_la-rempi_clock.o >/dev/null 2>&1
     805     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_encoder.lo -MD -MP -MF .deps/librempi_la-rempi_encoder.Tpo -c rempi_encoder.cpp
               -fPIC -DPIC -o .libs/librempi_la-rempi_encoder.o
     806     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_encoder_basic.lo -MD -MP -MF .deps/librempi_la-rempi_encoder_basic.Tpo -c rempi
             _encoder_basic.cpp  -fPIC -DPIC -o .libs/librempi_la-rempi_encoder_basic.o
  >> 807     make[2]: *** [Makefile:762: librempi_la-rempi_recorder.lo] Error 1
     808     make[2]: *** Waiting for unfinished jobs....
     809     libtool: compile:  /spack/opt/spack/linux-ubuntu22.04-x86_64/gcc-11.2.0/mpich-4.0.2-gpzf75yhrekurnz3v63fmevfjez2x2q2/bin/mpic++ -DHAVE_CONFIG_H -I. -I.. -DREMPI_L
             ITE -fPIC -g -Wall -pthread -std=c++0x -I/include/ -I -g -O2 -MT librempi_la-rempi_sig_handler.lo -MD -MP -MF .deps/librempi_la-rempi_sig_handler.Tpo -c rempi_sig
             _handler.cpp -o librempi_la-rempi_sig_handler.o >/dev/null 2>&1
     810     rempi_re.cpp: In member function 'virtual MPI_Fint rempi_re::re_request_c2f(MPI_Request)':
     811     rempi_re.cpp:515:10: warning: 'ret' is used uninitialized [-Wuninitialized]
     812       515 |   return ret;
     813           |          ^~~

     ...

     995     mv -f .deps/librempi_la-rempi_encoder_basic.Tpo .deps/librempi_la-rempi_encoder_basic.Plo
     996     mv -f .deps/librempi_la-rempi_pmpi.Tpo .deps/librempi_la-rempi_pmpi.Plo
     997     mv -f .deps/librempi_la-rempi_clock_delta_compression.Tpo .deps/librempi_la-rempi_clock_delta_compression.Plo
     998     mv -f .deps/librempi_la-rempi_encoder.Tpo .deps/librempi_la-rempi_encoder.Plo
     999     mv -f .deps/librempi_la-rempi_request_mg.Tpo .deps/librempi_la-rempi_request_mg.Plo
     1000    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-rempi-1.1.0-eeyr3icczf6v2agskpafphuqe5txl7uz/spack-src/src'
  >> 1001    make[1]: *** [Makefile:415: all-recursive] Error 1
     1002    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-rempi-1.1.0-eeyr3icczf6v2agskpafphuqe5txl7uz/spack-src'
  >> 1003    make: *** [Makefile:346: all] Error 2
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (b32cb5765c598e59f2f91d99807b4331f4eba569)
* **Python:** 3.10.4
* **Platform:** linux-ubuntu22.04-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9332528/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/9332529/spack-build-env.txt)

FYI @wspear @lee218llnl 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/rempi/include-string.patch
+++ b/var/spack/repos/builtin/packages/rempi/include-string.patch
@@ -1,0 +1,11 @@
+diff -ruN spack-src/src/rempi_message_manager.h spack-src-patched/src/rempi_message_manager.h
+--- spack-src/src/rempi_message_manager.h	2018-06-22 11:07:41.000000000 -0700
++++ spack-src-patched/src/rempi_message_manager.h	2023-01-17 07:24:04.252076735 -0800
+@@ -26,6 +26,7 @@
+ 
+ #include <unordered_map>
+ #include <list>
++#include <string>
+ 
+ #include "mpi.h"
+ 

--- a/var/spack/repos/builtin/packages/rempi/package.py
+++ b/var/spack/repos/builtin/packages/rempi/package.py
@@ -23,12 +23,13 @@ class Rempi(AutotoolsPackage):
     depends_on("libtool", type="build")
     depends_on("libpciaccess", type="link")
 
+    patch("include-string.patch", when="@1.1.0")
+
     def flag_handler(self, name, flags):
-        iflags = []
         if name == "cflags":
             if self.spec.satisfies("%oneapi@2022.2.0:"):
-                iflags.append("-Wno-error=implicit-function-declaration")
-        return (iflags, None, None)
+                flags.append("-Wno-error=implicit-function-declaration")
+        return (flags, None, None)
 
     def setup_build_environment(self, env):
         if self.spec.satisfies("%cce"):

